core/src/main/java/org/apache/iceberg/util/FilterIterator.java

```diff
@@ -34,21 +34,21 @@
 public abstract class FilterIterator<T> implements CloseableIterator<T> {
   private final Iterator<T> items;
   private boolean closed;
-  private boolean hasNext;
+  private boolean nextReady;
```

Comment:
This fixes the filter with reused containers, like `GenericRecord`. The `advance` call in `next` would replace values in a reused row, which would in effect return the next matching row.